### PR TITLE
[9.x] Support step argument for Collection::range method

### PR DESCRIPTION
**Pull request by trovster**
The `Collection::range()` method uses the default PHP `range()` method, but with no way of adding the step amount. This PR exposes that argument, with the same default as before. This also supports `LazyCollection::range()` and updates the interface. Tests provided.

### CODE REVIEW

File: tests/Support/SupportCollectionTest.php [hunk 1]
```diff
@@ -2421,6 +2422,21 @@ public function testRangeMethod($collection)
...
+
+        $this->assertSame(
+            [-5, -3, -1, 1, 3, 5],
+            $collection::range(-5, 5, 2)->all()
+        );
+
+        $this->assertSame(
+            [0, 5],
+            $collection::range(0, 5, 5)->all()
+        );
```
Comment (windu2b) [inline tests/Support/SupportCollectionTest.php:2439]: You test when `$from > $to`, but neither when `$from == $to` nor when `$from < $to`

**Comment by driesvints** [comment]
Ping @JosephSilber

File: src/Illuminate/Collections/Enumerable.php [hunk 2]
```diff
@@ -32,9 +32,10 @@ public static function times($number, callable $callback = null);
      *
      * @param  int  $from
      * @param  int  $to
+     * @param  int  $step
      * @return static
      */
-    public static function range($from, $to);
+    public static function range($from, $to, $step = 1);
```
Comment (GrahamCampbell) [inline src/Illuminate/Collections/Enumerable.php:38]: breaking change

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Collections/Enumerable.php (hunk 2 above)
Comment (driesvints) [reply src/Illuminate/Collections/Enumerable.php:38]: @GrahamCampbell  that's why this is targetting master

**Review by JosephSilber — COMMENTED** [review]
1. What happens when `$step` is passed in as `0`?

    PHP's built-in `range` currently returns `false`, and emits a warning. The lazy implementation here would simply go into an infinite loop.

2. What happens when `$step` is a negative number?

    [The docs](https://www.php.net/manual/en/function.range.php) say:

    > `step` should be given as a positive number

    However, passing it a negative number actually just treats it as a positive number, which just feels wrong. And the current lazy implementation behaves totally differently.

    I think it would make sense to have both implementations throw initially if `$step <= 0`.

3. It would be beneficial to add a test that ensures the step doesn't overshoot.

    So `range(0, 7, 5)` returns `[0, 5]`, not `[0, 5, 10]`.

**Comment by driesvints** [comment]
@trovster gonna close this one now. Feel free to open up a new one with replies to @JosephSilber's remarks when you find the time.

**Pull request closed without merging**